### PR TITLE
refactor: move logic from Request classes to new service layer

### PR DESCRIPTION
**Pull request by jhaeu**
### Pull Request Checklist

- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [ ] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [ ] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [ ] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [ ] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [ ] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [ ] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [ ] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes # .

### Information about the changes
- Key functionality added: New service classes with injected EndpointProperties. The logic to create internal requests is moved from APIRequest and its subclasses to the new services. 
- Reason for change: EnspointProperties were only injected to the controllers. We could have passed these properties to the old methods in APIRequest subclasses, or extend the APIRequest classes and set the endpointProperties in the controllers or separate the data of the APIRequest classes and the logic to create internal requests based on their data. I chose the last approach to get a better separation of concerns.

### Examples and reasons for differences between live ORS routes, and those generated from this pull request
-

### Required changes to ors config (if applicable)
-

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines

### CODE REVIEW

File: ors-api/src/main/java/org/heigit/ors/api/services/RoutingService.java [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Stream;
+
+
+
+@Service
+public class RoutingService extends AbstractApiService {
+
+    private final EndpointsProperties endpointsProperties;
```
Comment (aoles) [inline ors-api/src/main/java/org/heigit/ors/api/services/RoutingService.java]: As all `*Service` classes seem to share a common `private final EndpointsProperties endpointsProperties;`field I would suggest moving it along with the corresponding getter to the parent `AbstractApiService` class.

File: ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java [hunk 2]
```diff
@@ -0,0 +1,424 @@
...
+import org.locationtech.jts.geom.Polygon;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
+public abstract class AbstractApiService {
+
+    abstract EndpointsProperties getEndpointsProperties();
+    abstract double getMaximumAvoidPolygonArea();
```
Comment (aoles) [inline ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java]: All `*Service` classes but `RoutingService` return 0 in `getMaximumAvoidPolygon*` methods. Therefore, you might consider defining this default implementation already here in the upstream abstract class and override it only when relevant.

File: ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java [hunk 3]
```diff
@@ -0,0 +1,424 @@
...
+                    return BordersExtractor.Avoid.ALL;
+                case CONTROLLED:
+                    return BordersExtractor.Avoid.CONTROLLED;
+                default:
+                    return BordersExtractor.Avoid.NONE;
+            }
+        }
+        return null;
+    }
+
```
Comment (aoles) [inline ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java]: Although most of the added code is in fact just code moved from the original classes, I've noticed that in this process some double blank lines have been introduced to separate method bodies. Was this change intentional?

File: ors-api/src/test/java/org/heigit/ors/api/services/APIServiceTest.java [hunk 4]
```diff
@@ -19,56 +21,72 @@
 
 import static org.junit.jupiter.api.Assertions.*;
 
-class APIRequestTest {
-    APIRequest request;
+class APIServiceTest {
+    AbstractApiService apiService;
 
+    static class ApiService extends AbstractApiService {
```
Comment (aoles) [inline ors-api/src/test/java/org/heigit/ors/api/services/APIServiceTest.java]: Might become obsolete once the abstract methods in `AbstractApiService` become default implementations as suggested in my other comments.

**Review by aoles — CHANGES_REQUESTED** [review]
Awesome, thanks a lot @jhaeu for your contribution!

This looks quite good already. Just a few small things to consider, see my inline comments.

On a general note, please go through all of the modified classes and make sure to remove any unused imports 😇

File: ors-api/src/main/java/org/heigit/ors/api/services/RoutingService.java (hunk 1 above)
Comment (jhaeu) [reply ors-api/src/main/java/org/heigit/ors/api/services/RoutingService.java]: will change

File: ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java (hunk 3 above)
Comment (jhaeu) [reply ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java]: No, I removed many multiple blank lines, but maybe not all... I will reformat all changed classes.

File: ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java (hunk 2 above)
Comment (jhaeu) [reply ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java]: changed

File: ors-api/src/test/java/org/heigit/ors/api/services/APIServiceTest.java (hunk 4 above)
Comment (jhaeu) [reply ors-api/src/test/java/org/heigit/ors/api/services/APIServiceTest.java]: changed

File: ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java (hunk 3 above)
Comment (jhaeu) [reply ors-api/src/main/java/org/heigit/ors/api/services/AbstractApiService.java]: Removed some double blank lines, but decided not to reformat the classes NOW - Amandus mentioned, that a big reformatting of the whole project in an extra branch is planned. BTW in our .editorconfig we seem to allow double blank lines if I'm not wrong?

**Review by takb — APPROVED** [review]
Looks all good to me now. Excellent! @aoles have your requests been addressed?

**Review by aoles — APPROVED** [review]
Thanks a lot @jhaeu for accommodating my suggestions, looks much cleaner now :)

I think we are good to go 🚀

Cheers!
